File: src/main/java/org/app/infrastructure/local/InMemoryUserRepository.java
package org.app.infrastructure.local;

import jakarta.enterprise.context.ApplicationScoped;
import org.app.domain.IUserRepository;
import org.app.domain.User;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@ApplicationScoped
public class InMemoryUserRepository implements IUserRepository {
    private List<UserData> users;

    public InMemoryUserRepository() {
        this.users = new ArrayList<>();
    }

    @Override
    public void register(User user) {
        UserData userData = UserData.toData(user);
        this.users.add(userData);
    }

    private Stream<UserData> filterUserId(List<UserData> users, String userId){
        return users.stream().filter(user -> user.getId().equals(userId));
    }

    @Override
    public List<User> find(UUID userId) {
        List<UserData> userData = filterUserId(this.users, userId.toString()).collect(Collectors.toList());
        return userData.stream().map(UserData::toDomain).collect(Collectors.toList());
    }

    @Override
    public List<User> find(List<UUID> userIds) {
        return users.stream().filter(user -> userIds.stream().anyMatch(userId -> user.getId().equals(userId.toString()))).map(UserData::toDomain).collect(Collectors.toList());
    }

    private List<UserData> filterUserUsername(List<UserData> users, String username){
        return users.stream().filter(user -> user.getUsername().equals(username)).collect(Collectors.toList());
    }

    @Override
    public List<User> findByUsername(String username) {
        List<UserData> userData = filterUserUsername(this.users, username);
        return userData.stream().map(UserData::toDomain).collect(Collectors.toList());
    }

    @Override
    public void update(List<User> users) {
        List<UserData> userData = users.stream().map(UserData::toData).collect(Collectors.toList());
        this.users = this.users.stream().map(user -> filterUserId(userData, user.getId()).findFirst().orElse(user)).collect(Collectors.toList());
    }

    public void removeUsers(List<User> users) {
        List<UserData> userData = users.stream().map(UserData::toData).collect(Collectors.toList());
        this.users = this.users.stream().filter(user -> filterUserId(userData, user.getId()).findFirst().isEmpty()).collect(Collectors.toList());
    }
}

File: go/app/main/infrastructure/local/in_memory_user_repository.go
package local

import (
    "github.com/google/uuid"
    "mutualAttentionSystem/app/main/domain"
)

type InMemoryUserRepository struct {
    users []*UserData
}

func NewInMemoryUserRepository() *InMemoryUserRepository {
    return &InMemoryUserRepository{}
}

func filterUserId(users []*UserData, id string) *UserData {
    for _, user := range users {
        if user.ID == id {
            return user
        }
    }
    return nil
}

func (repository *InMemoryUserRepository) Find(id uuid.UUID) *domain.User {
    userData := filterUserId(repository.users, id.String())
    if userData == nil {
        return nil
    }
    return toUserDomain(userData)
}

func (repository *InMemoryUserRepository) FindByIds(ids []uuid.UUID) []*domain.User {
    var users []*domain.User
    for _, id := range ids {
        userData := filterUserId(repository.users, id.String())
        if userData == nil {
            continue
        }

        users = append(users, toUserDomain(userData))
    }
    return users
}

func filterUsername(users []*UserData, username string) *UserData {
    for _, user := range users {
        if user.Username == username {
            return user
        }
    }
    return nil
}

func (repository *InMemoryUserRepository) FindByUsername(username string) *domain.User {
    userData := filterUsername(repository.users, username)
    if userData == nil {
        return nil
    }
    return toUserDomain(userData)
}

func (repository *InMemoryUserRepository) Save(user *domain.User) {
    userData := toUserData(user)
    repository.users = append(repository.users, userData)
}
